jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java

```diff
@@ -256,75 +267,45 @@ public boolean isContainedIn(Resource r)
-         * If the URI is different from the Path.toUri() then
-         * we will just use the original URI to construct the
-         * alias reference Path.
-         *
-         * We use the method `toUri(Path)` here, instead of
-         * Path.toUri() to ensure that the path contains
-         * a trailing slash if it's a directory, (something
-         * not all FileSystems seem to support)
-         */
-        if (!URIUtil.equalsIgnoreEncodings(uri, toUri(path)))
+        if (targetPath == null)
         {
-            try
-            {
-                // Use normalized path to get past navigational references like "/bar/../foo/test.txt"
-                Path ref = Paths.get(uri.normalize());
-                return ref.toRealPath();
-            }
-            catch (IOException ioe)
+            targetPath = resolveTargetPath();
+            if (Files.exists(targetPath))
             {
-                // If the toRealPath() call fails, then let
-                // the alias checking routines continue on
-                // to other techniques.
-                LOG.trace("IGNORED", ioe);
+                if (!isSameName(path, targetPath))
+                    alias = true;
+                else if (!URIUtil.equalsIgnoreEncodings(uri, toUri(path)))
+                {
+                    /* Catch situation where the Path class has already normalized
+                     * the URI e.g. input path "aa./foo.txt"
+                     * from an #resolve(String) is normalized away during
+                     * the creation of a Path object reference.
+                     * If the URI is different from the Path.toUri() then
+                     * we will just use the original URI to construct the
+                     * alias reference Path.
+                     *
+                     * We use the method `toUri(Path)` here, instead of
+                     * Path.toUri() to ensure that the path contains
+                     * a trailing slash if it's a directory, (something
+                     * not all FileSystems seem to support)
+                     */
+                    alias = true;
+                }
             }
         }
+    }
 
-        if (!abs.isAbsolute())
-            abs = path.toAbsolutePath();
-
-        // Any normalization difference means it's an alias,
-        // and we don't want to bother further to follow
-        // symlinks as it's an alias anyway.
-        Path normal = path.normalize();
-        if (!isSameName(abs, normal))
-            return normal;
-
+    private Path resolveTargetPath()
+    {
         try
         {
-            if (Files.isSymbolicLink(path))
-                return path.getParent().resolve(Files.readSymbolicLink(path));
-            if (Files.exists(path))
-            {
-                Path real = abs.toRealPath();
-                if (!isSameName(abs, real))
-                    return real;
-            }
+            return path.normalize().toRealPath();
```

Comment:
`normalize()` shouldn't be necessary here.
The `toRealPath()` should do that for you.
Got an example of where it doesn't?

Reply:
So if the Resource does not exist, and the targetPath also does not exist (or can be determined).
Is it safe to have this `resource.isAlias()` return false?

What if the resource gets created later?

What if the resource is a symlink that doesn't point to anything? (this would be an exception in `.toRealPath()` as well)
And then it does point to something after the resource is created?

Reply:
> Got an example of where it doesn't?

If you give a path with un-normalized dot dot segments it will report file does not exist and `toRealPath()` will throw, but if you normalize first it resolves correctly to the existing file.
I encountered this during the testing of this change.

> This should not return the path.normalize() like you do in line 319, as that's a bad reference.

The reason I return `path.normalize()` on line 319 is so that a path like `/foo/bar/../something` would resolve as an alias to `/foo/something` without having to exist. I don't think paths with unresolved `.` or `..` segments should ever be classified as not an alias.

> What if the resource gets created later?

This is a good point, perhaps anything that does not exist should be reported as being an alias.